### ISSUE TITLE
11.15.2 react-native TS errors

### ISSUE THREAD

**Issue by zoominGiladSher** [i18next/react-i18next#1434]
## 🐛 Bug Report

Hello, we've been experiencing a problem with the new version that was released today. It fails our builds with the following error:
```
../node_modules/react-i18next/ts4.1/index.d.ts(11,27): error TS2307: Cannot find module 'react-native' or its corresponding type declarations.
../node_modules/react-i18next/ts4.1/index.d.ts(242,3): error TS2706: Required type parameters may not follow optional type parameters.
```

## To Reproduce

I'm not sure if it's our dependencies or a bug in the new types that were added for the version. For now, we locked the version on 11.15.1 and it solved the issue

## Expected behavior
New version doesn't  break builds

## Your Environment
node 16
typescript 4.5.4
react-i18next@11.15.2
i18next@20.4.0
*os*: Linux bitbucket pipeline
- also fails locally on MacOS 12.0.1 (21A559)

**Comment by adrai** [comment]
@reichhartd 
https://github.com/i18next/react-i18next/issues/1418
https://github.com/i18next/react-i18next/pull/1419
//cc @pedrodurek

**Comment by reichhartd** [comment]
@zoominGiladSher Can you provide a minimal [CodeSandbox](https://codesandbox.io/) reproduction where the error occurs?
Unfortunately, I can't reproduce it. The latest version works for all my projects.

**Comment by zoominGiladSher** [comment]
Hi @reichhartd , I can't seem to reproduce it on code sandbox, so I assume it's some compatibility issue with our versions of the packages. Since it's a huge project I can't really get it all into code sandbox. If you can't reproduce it you may close the issue and we'll try to handle it on our own via package upgrades in the future.

Thank you for taking the time to look at this!

**Comment by PerfectionVR** [comment]
https://codesandbox.io/s/wild-bush-dwlif?file=/src/App.tsx I tried to make a minimum example, but I believe that codesandbox auto resolves react-native.

Right now, we can't use this version of react-i18next without importing react-native, which is something we want to avoid in a non-react native package and is thus a breaking change for us...

**Comment by adrai** [comment]
If someone can reproduce this in an example github repository, this would help.

**Comment by bfeigin** [comment]
Also experiencing this issue. With a bit more of the error log pasted, it seems like the ts4.1/index.d.ts is expecting react-native to always be available. 
```
node_modules/react-i18next/ts4.1/index.d.ts:11:27 - error TS2307: Cannot find module 'react-native' or its corresponding type declarations.

11 import { TextProps } from 'react-native';
                             ~~~~~~~~~~~~~~

node_modules/react-i18next/ts4.1/index.d.ts:242:3 - error TS2706: Required type parameters may not follow optional type parameters.
242   E extends React.HTMLProps<HTMLDivElement> | TextProps
```

Looking at the diff between 15.1 => 15.2 it seems like `@types/react-native` was pulled in as a [dev dependency]( https://github.com/i18next/react-i18next/compare/v11.15.1...v11.15.2#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R52) but appears to be required for the [index.d.ts file](https://github.com/i18next/react-i18next/compare/v11.15.1...v11.15.2#diff-61782f9118a033e04ba53dd09fb1e38d10dc39d2dfb9f8474f3dad935f16ea21R11)

**Comment by adrai** [comment]
@reichhartd @pedrodurek is the react-native dependency intensional? Don't like it to be mandatory for non-ts users.

**Comment by reichhartd** [comment]
@adrai Yes `@types/react-native` is intensional. But it is only the TypeScript types and not the complete `react-native` package.
My suggestion would be to drag `@types/react-native` from the dev dependencies into the dependencies.

**Comment by reichhartd** [comment]
@adrai Yes @types/react-native is intensional. But it is only the TypeScript types and not the complete react-native package.
My suggestion would be to drag @types/react-native from the dev dependencies into the dependencies.

**Comment by adrai** [comment]
Why should a NON-typescript user install `@types/react-native` ?
Why should a NON-react-native user install `@types/react-native` ?
Is this good practice? - This does not sound correct to me.
Currently, I'm thinking to revert that change.

**Comment by reichhartd** [comment]
Because you can't install npm packages conditionally.

If you would use `@types/react-dom` types in your type definitions, I can ask the counter question: Why should a react-native user install `@types/react-dom` ?

You can also copy `TextProps` from `@types/react-native`, then you don't need the dependency. But in this case you have to make sure that `TextProps` is always up-to-date with `@types/react-native`.
One could also replace `TextProps` with `any`, but in this case the type definitions are then less secure.

You can also revert the change, but in this case the TypeScript types for react-native users are simply wrong.

There is no perfect solution.

**Comment by adrai** [comment]
I'm just asking "stupid" questions, because I'm not a TypeScript user...

Regarding `@types/react-dom`, I don't see any dependency declaration of the types in the package.json... maybe this works differently? Does this dependency need to be installed on user level? So also `@types/react-native` can be installed on user level?

Just asking...

**Comment by reichhartd** [comment]
There are no stupid questions 🙂

That's right. `@types/react-dom` is not installed and used. Currently only the types of `i18next` (which is delivered directly with the package) and `react` are used. For `@types/react` we can assume that all our TypeScript users have it installed.

With `@types/react-dom` I just wanted to make the point that if you support multiple libraries like `react-dom` ("react") and `react-native`, you could come to a point where you have to build exceptions for one of them. Let it be `react-dom` or `react-native`.

Yes, you could install `@types/react-native` at a user level. Then only TypeScript users would have to do this. However, this dependency would also have to be installed by `react-dom` TypeScript users. 
```
  E extends React.HTMLProps<HTMLDivElement> | TextProps = React.HTMLProps<HTMLDivElement>
```
https://github.com/i18next/react-i18next/blob/master/ts4.1/index.d.ts#L262

Think of types as variables. Even if `react-dom` users don't need the type `TextProps`, TypeScript checks the generic `E` to validate if it extends `React.HTMLProps<HTMLDivElement>` or `TextProps` correctly. If `TextProps` is not found, the TypeScript compiler throws an error.

To be honest, I'm not sure what the best practice is here either. I have listed all the possibilities I could think of in this answer and in the previous one.
I don't think this is a special case and I will look at other TypeScript libraries today to see how they handle it.
If @pedrodurek could think of other possibilities, I would be happy to discuss them.

**Comment by reichhartd** [comment]
# tl;dr:

Let's change it to `E = React.HTMLProps<HTMLDivElement>`

# Long answer

The problem with `react` projects is not only that an additional package `@types/react-native` is installed, but also that the global definitions of `@types/react-native` interfere with the DOM standard library `lib.dom.d.ts`. [`@types/styled-components` had problems with this for over 1.5 years.](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/49914)

One would need to import `TextProps` from `@types/react-native/index.d.ts` without this `@types/react-native/globals.d.ts` with the [Triple-Slash Directives](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html).
Since `@types/react-native` creates several problems with regular `react` projects, I think the best solution is to make the generic `E` less type-safe by having it accept style objects from `react` and `react-native`.

I have tested the above solution with a freshly set up `react` and `react-native` project.
@adrai If you agree with me, I will create a pull request.

**Comment by adrai** [comment]
sounds good
